src/search/FindInFiles.js

```diff
@@ -734,67 +742,81 @@ define(function (require, exports, module) {
          * Remove existing search results that match the given entry's path
          * @param {(File|Directory)} entry
          */
-        function _removeSearchResultsForEntry(entry) {
-            Object.keys(searchModel.results).forEach(function (fullPath) {
-                if (fullPath === entry.fullPath ||
-                        (entry.isDirectory && fullPath.indexOf(entry.fullPath) === 0)) {
-                    // node search : inform node that the file is removed
-                    filesRemoved([fullPath]);
-                    if (findOrReplaceInProgress) {
-                        searchModel.removeResults(fullPath);
-                        resultsChanged = true;
+        function _removeSearchResultsForEntries(entries) {
+            var fullPaths = [];
+            entries.forEach(function (entry) {
+                Object.keys(searchModel.results).forEach(function (fullPath) {
+                    if (fullPath === entry.fullPath ||
+                            (entry.isDirectory && fullPath.indexOf(entry.fullPath) === 0)) {
+                        // node search : inform node that the file is removed
+                        fullPaths.push(fullPath);
+                        if (findOrReplaceInProgress) {
+                            searchModel.removeResults(fullPath);
+                            resultsChanged = true;
+                        }
                     }
-                }
+                });
             });
+            // this should be called once with a large array instead of numerous calls with single items
+            filesRemoved(fullPaths);
         }
 
         /*
          * Add new search results for this entry and all of its children
          * @param {(File|Directory)} entry
          * @return {jQuery.Promise} Resolves when the results have been added
          */
-        function _addSearchResultsForEntry(entry) {
-            var addedFiles = [],
-                addedFilePaths = [],
-                deferred = new $.Deferred();
-
-            // gather up added files
-            var visitor = function (child) {
-                // Replicate filtering that getAllFiles() does
-                if (ProjectManager.shouldShow(child)) {
-                    if (child.isFile && _isReadableText(child.name)) {
-                        // Re-check the filtering that the initial search applied
-                        if (_inSearchScope(child)) {
-                            addedFiles.push(child);
-                            addedFilePaths.push(child.fullPath);
+        function _addSearchResultsForEntries(entries) {
+            var fullPaths = [];
```

Comment:
Well, this array is not being persisted besides the function execution lifetime and I wouldn't expect there to be a problem even if you have thousands of entries in it. On the other hand, without it, trying to send thousands of messages to a node process will definitely crash it (or socket server). In other words, I don't think memory usage is something to worry about.